<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="phylogram">
<title>phylogram: an R package for phylogenetic analysis with dendrograms • phylogram</title>
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="/apple-touch-icon.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://docs.ropensci.org/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="phylogram: an R package for phylogenetic analysis with dendrograms">
<meta property="og:description" content="phylogram">
<meta property="og:image" content="https://docs.ropensci.org/phylogram/logo.png">
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Matomo --><link rel="stylesheet" type="text/css" href="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.css">
<script src="https://cdn.jsdelivr.net/npm/cookieconsent@3/build/cookieconsent.min.js" data-cfasync="false"></script><script src="https://ropensci.org/scripts/matomo.js"></script><noscript><p><img src="https://ropensci.matomo.cloud/matomo.php?idsite=1&amp;rec=1" style="border:0;" alt=""></p></noscript>
<!-- End Matomo Code -->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    <a href="https://ropensci.org" class="external-link"><img src="https://ropensci.org/img/icon_short_white.svg" id="hexlogo" alt="rOpenSci"></a>
    <a class="navbar-brand me-2" href="../index.html">phylogram</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">2.1.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"></ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/phylogram-vignette.html">phylogram: an R package for phylogenetic analysis with dendrograms</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="http://github.com/ropensci/phylogram/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>phylogram: an R package for phylogenetic analysis with dendrograms</h1>
                        <h4 data-toc-skip class="author">Shaun P. Wilkinson <em><sup>1,2</sup></em> and Simon K. Davy <em><sup>1</sup></em>
</h4>
            
      
      <small class="dont-index">Source: <a href="http://github.com/ropensci/phylogram/blob/HEAD/vignettes/phylogram-vignette.Rmd" class="external-link"><code>vignettes/phylogram-vignette.Rmd</code></a></small>
      <div class="d-none name"><code>phylogram-vignette.Rmd</code></div>
    </div>

    
    
<hr>
<p><em><sup>1</sup></em> School of Biological Sciences, Victoria University of Wellington, P.O. Box 600, Wellington, New Zealand.</p>
<p><em><sup>2</sup></em> Correspondence author. E-mail: shaunpwilkinson@gmail.com</p>
<p> </p>
<div class="section level2">
<h2 id="abstract">Abstract<a class="anchor" aria-label="anchor" href="#abstract"></a>
</h2>
<p>The <strong>phylogram</strong> R package is a tool for for developing phylogenetic trees as deeply-nested lists known as “dendrogram” objects. It provides functions for conversion between dendrograms and “phylo” class objects, as well as several tools for command-line tree manipulation and import/export via Newick parenthetic text. This improves accessibility to the comprehensive range of object-specific analytical and tree-visualization functions found across a wide array of bioinformatic R packages. The <strong>phylogram</strong> package is released under the GPL-3 license, and is available for download from CRAN <a href="https://CRAN.R-project.org/package=phylogram" class="external-link uri">https://CRAN.R-project.org/package=phylogram</a> and GitHub <a href="https://github.com/ropensci/phylogram" class="external-link uri">https://github.com/ropensci/phylogram</a>.</p>
<p> </p>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The R environment continues to gain popularity as a platform for bioinformatic analysis, due to the reproducible code-based workflow and the many powerful analytical tools available in a suite of open-source packages such as <strong>ape</strong> <span class="citation">(Paradis, Claude, and Strimmer 2004)</span>, <strong>phangorn</strong> <span class="citation">(Schliep 2011)</span> and <strong>Phytools</strong> <span class="citation">(Revell 2012)</span>. These packages typically employ a tree structure known as the “phylo” object, whose primary element is an integer matrix with one row for each edge in the graph, and two columns giving the indices of the connecting nodes. This is a versatile and memory-efficient representation suitable for most applications encountered by evolutionary biologists, and hence a comprehensive array of tools has been developed for parsing, manipulating, analyzing and visualizing trees in matrix format.</p>
<p>An alternative tree structure is the “dendrogram” object, generated using the <code>as.dendrogram</code> function in the <strong>stats</strong> package <span class="citation">(R Core Team 2017)</span>. Rather than a matrix of edges, a dendrogram is a hierarchical list. These ‘lists of lists’ can be deeply nested, with the limit depending on the C stack size (settable <em>via</em> <code>options("expressions")</code>). While less memory-efficient than matrix-based tree objects, a useful feature of the nested-list representation is its modularity, whereby the sub tree of a tree is itself a tree - a dendrogram within a dendrogram. This means that dendrogram objects are subsettable in the same way that standard lists are, which in addition to the inbuilt editing functions such as <code>cut</code> and <code>merge</code>, facilitates intuitive command-line tree manipulation. An especially powerful feature of this object type is that tree editing operations can be carried out recursively using fast inbuilt functions in the “apply” family such as <code>dendrapply</code> and <code>lapply</code>.</p>
<p>Each node of a dendrogram object has the following mandatory attributes:</p>
<ul>
<li>“height” the position of the node along the vertical axis (assuming the graph is orientated vertically)</li>
<li>“midpoint” the horizontal distance of the node from the left-most member of the sub tree (where the horizontal distance between adjacent leaves is 1 unit)</li>
<li>“members” the number of terminal leaf nodes belonging to the node</li>
<li>“class” all nodes have the class attribute “dendrogram”</li>
</ul>
<p>Rather than lists, terminal leaf nodes are length-1 integer vectors whose values correspond to the indices of the members in the set. The “members” attributes of leaf nodes is always 1, the “midpoint” attribute is 0, and they have two additional attributes:</p>
<ul>
<li>“leaf” TRUE for terminal nodes (NULL otherwise)</li>
<li>“label” an optional character string giving the name of the taxon or group</li>
</ul>
<p>Aside from those listed above, users may attach other objects as attributes to the dendrogram nodes. For example, “label” attributes can be attached to inner nodes, and users can specify plotting parameters for each node by setting the attributes “nodePar” and “edgePar”. Any number of non-standard objects can also be attached as node attributes, which can be useful for storing additional metadata such as bootstrap values and taxonomic information.</p>
<p>The flexibility, modularity and intuitive structure of dendrogram objects are appealing to many users, particularly where highly dynamic tree structures are required for applications such as machine learning clustering and classification. There is also a large and growing number of resources for manipulating and plotting dendrograms in contributed packages such as <strong>dendextend</strong> <span class="citation">(Galili 2015)</span>, and hence functions enabling bi-directional conversion between “dendrogram” and “phylo” class objects would expand the range of tools available for both object types. While conversion is currently possible using the “hclust” object as an intermediary, this object type does not support non-zero leaf node heights and hence is limited to ultrametric trees only.</p>
<p> </p>
</div>
<div class="section level2">
<h2 id="the-phylogram-package">The phylogram package<a class="anchor" aria-label="anchor" href="#the-phylogram-package"></a>
</h2>
<p>Here, we introduce <strong>phylogram</strong>, an R package for structuring evolutionary trees as deeply-nested lists and transforming trees between list- and matrix-type objects. The package also contains functions for importing and exporting dendrogram objects to and from parenthetic text, as well as several functions for manipulating trees in nested-list format. These functions are detailed below with examples of their utility.</p>
<p> </p>
<div class="section level3">
<h3 id="importing-and-exporting-trees">Importing and exporting trees<a class="anchor" aria-label="anchor" href="#importing-and-exporting-trees"></a>
</h3>
<p>The Newick parenthetic text format [a.k.a. the New Hampshire format; <span class="citation">Felsenstein et al. (1986)</span>] is a universal phylogenetic tree representation that is compatible with most tree-editing software. The <strong>phylogram</strong> function <code>read.dendrogram</code> wraps the the Newick parser <code>read.tree</code> from the <strong>ape</strong> package <span class="citation">(Paradis, Claude, and Strimmer 2004)</span>, and converts the intermediate “phylo” object to a dendrogram. This function supports weighted edges, labels with special meta-characters (enclosed in single quotation marks), comments (enclosed in square brackets; ignored by the parser), multifuricating nodes, and both rooted and unrooted trees. Inner-node labels are also supported, and are attached as “label” attributes to non-leaf nodes. Objects of class “dendrogram” can also be exported as Newick-style parenthetic text using the function <code>write.dendrogram</code>.</p>
<p> </p>
<div class="section level4">
<h4 id="example-1-import-and-export-a-tree-from-a-newick-string">Example 1: Import and export a tree from a Newick string<a class="anchor" aria-label="anchor" href="#example-1-import-and-export-a-tree-from-a-newick-string"></a>
</h4>
<p>Consider the simple example of a tree with three members named “A”, “B” and “C”, where “B” and “C” are more closely related to each other than either is to “A”. An unweighted Newick string for this tree would be <em>(A,(B,C));</em> This can be imported as a dendrogram object using the <code>read.dendrogram</code> function as follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/phylogram">phylogram</a></span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read.dendrogram.html">read.dendrogram</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">"(A,(B,C));"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, yaxt <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/unnamed-chunk-3-1.png" width="500px" style="display: block; margin: auto auto auto 0;"></p>
<p>    <strong>Figure 1:</strong> A simple dendrogram with three terminal leaf nodes</p>
<p> </p>
<p>The following command writes the object back to the console in Newick format without edge weights:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write.dendrogram.html">write.dendrogram</a></span><span class="op">(</span><span class="va">x</span>, edges <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "(A,(B,C));"</span></span></code></pre></div>
<p>The syntax is similar when reading and writing text files, except that the <code>text</code> argument is replaced by <code>file</code>, and a valid file path is passed to the function.</p>
<p> </p>
</div>
</div>
<div class="section level3">
<h3 id="converting-tree-objects">Converting tree objects<a class="anchor" aria-label="anchor" href="#converting-tree-objects"></a>
</h3>
<p>Dendrograms can be converted to “phylo” objects and <em>vice versa</em> using the <code>as.phylo.dendrogram</code> and <code>as.dendrogram.phylo</code> methods. Unlike functions that employ an “hclust” object as an intermediary (e.g. <code>as.dendrogram(as.hclust(phy))</code>), these methods retain all weighted edges and do not require trees to be ultrametric. This facilitates access to the comprehensive range of functions that are specific for either dendrograms or “phylo” objects in contributed packages such as <strong>dendextend</strong> <span class="citation">(Galili 2015)</span> and <strong>ape</strong> <span class="citation">(Paradis, Claude, and Strimmer 2004)</span>, respectively. Note that other packages may employ the same function names, and hence the method dispatched may depend on the order in the which packages are loaded. For this reason it may be safer to use the full function call (e.g. <code>phylogram::as.phylo.dendrogram(x)</code> and <code>phylogram::as.dendrogram.phylo(x)</code>) when using these methods.</p>
<p> </p>
<div class="section level4">
<h4 id="example-2-convert-a-phylo-object-to-a-dendrogram">Example 2: Convert a “phylo” object to a dendrogram<a class="anchor" aria-label="anchor" href="#example-2-convert-a-phylo-object-to-a-dendrogram"></a>
</h4>
<p>A common application requiring conversion between “phylo” and “dendrogram” objects involves plotting tanglegrams to visualize incongruence between two phylogenetic trees. The <strong>dendextend</strong> package <span class="citation">(Galili 2015)</span> features the function <code>tanglegram</code> for versatile plotting of two distinct trees and indicating the discordant nodes using a series of non-parallel edges. However this function does not currently support non-ultrametric “phylo” objects. In this example, two weighted neighbor-joining trees are generated from the left and right sections of the woodmouse alignment from the <strong>ape</strong> package <span class="citation">(Paradis, Claude, and Strimmer 2004)</span>, and converted to dendrograms for visual comparison using <code><a href="http://talgalili.github.io/dendextend/reference/tanglegram.html" class="external-link">dendextend::tanglegram</a></code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">woodmouse</span><span class="op">)</span></span>
<span><span class="co">## generate distance matrices for each section of the alignment</span></span>
<span><span class="va">dist1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/dist.dna.html" class="external-link">dist.dna</a></span><span class="op">(</span><span class="va">woodmouse</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">482</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">dist2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/dist.dna.html" class="external-link">dist.dna</a></span><span class="op">(</span><span class="va">woodmouse</span><span class="op">[</span>, <span class="fl">483</span><span class="op">:</span><span class="fl">965</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">## build neighbor-joining trees</span></span>
<span><span class="va">phy1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/nj.html" class="external-link">nj</a></span><span class="op">(</span><span class="va">dist1</span><span class="op">)</span></span>
<span><span class="va">phy2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/nj.html" class="external-link">nj</a></span><span class="op">(</span><span class="va">dist2</span><span class="op">)</span></span>
<span><span class="co">## root with No0912S as outgroup</span></span>
<span><span class="va">phy1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">root</a></span><span class="op">(</span><span class="va">phy1</span>, <span class="st">"No0912S"</span><span class="op">)</span></span>
<span><span class="va">phy2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/root.html" class="external-link">root</a></span><span class="op">(</span><span class="va">phy2</span>, <span class="st">"No0912S"</span><span class="op">)</span></span>
<span><span class="co">## convert phylo objects to dendrograms</span></span>
<span><span class="va">dnd1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html" class="external-link">as.dendrogram</a></span><span class="op">(</span><span class="va">phy1</span><span class="op">)</span></span>
<span><span class="va">dnd2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html" class="external-link">as.dendrogram</a></span><span class="op">(</span><span class="va">phy2</span><span class="op">)</span></span>
<span><span class="co">## rearrange in ladderized fashion</span></span>
<span><span class="va">dnd1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ladder.html">ladder</a></span><span class="op">(</span><span class="va">dnd1</span><span class="op">)</span></span>
<span><span class="va">dnd2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ladder.html">ladder</a></span><span class="op">(</span><span class="va">dnd2</span><span class="op">)</span></span>
<span><span class="co">## plot the tanglegram</span></span>
<span><span class="va">dndlist</span> <span class="op">&lt;-</span> <span class="fu">dendextend</span><span class="fu">::</span><span class="fu"><a href="http://talgalili.github.io/dendextend/reference/dendlist.html" class="external-link">dendlist</a></span><span class="op">(</span><span class="va">dnd1</span>, <span class="va">dnd2</span><span class="op">)</span></span>
<span><span class="fu">dendextend</span><span class="fu">::</span><span class="fu"><a href="http://talgalili.github.io/dendextend/reference/tanglegram.html" class="external-link">tanglegram</a></span><span class="op">(</span><span class="va">dndlist</span>, fast <span class="op">=</span> <span class="cn">TRUE</span>, margin_inner <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/unnamed-chunk-5-1.png" width="1000px" style="display: block; margin: auto auto auto 0;"></p>
<p>    <strong>Figure 2:</strong> Tanglegram showing incongruence between the left- and right-hand sections of the woodmouse alignment.</p>
<p> </p>
</div>
</div>
<div class="section level3">
<h3 id="tree-editingmanipulation">Tree editing/manipulation<a class="anchor" aria-label="anchor" href="#tree-editingmanipulation"></a>
</h3>
<p>The <strong>phylogram</strong> package features several additional functions to facilitate some of the more common manipulation operations. Leaf nodes and internal branching nodes can be removed using the function <code>prune</code>, which identifies and recursively deletes nodes based on pattern matching of “label” attributes. This is slower than the <strong>ape</strong> function <code>drop.tip</code>, but offers the benefits of versatile string matching using regular expressions, and the ability to remove inner nodes (and by extension all of their subnodes) that feature matching “label” attributes. To aid visualization, the function <code>ladder</code> rearranges the tree, sorting nodes by the number of members (analogous to the <code>ladderize</code> function in the <strong>ape</strong> package). Another function aiding in tree visualization is <code>as.cladogram</code>, which resets the “height” attributes of all terminal leaf nodes to zero and progressively resets the heights of the inner nodes by single incremental units in a bottom-up fashion. The function <code>reposition</code> scales the heights of all nodes in a tree by a given constant (passed <em>via</em> the argument <code>shift</code>), and features the option to reset all node heights so that height of the farthest terminal leaf node from the root is zero (by specifying <code>shift = "reset"</code>). The function <code>remidpoint</code> recursively corrects all “midpoint”, “members” and “leaf” attributes following manual editing of a tree or while converting a nested list to a “dendrogram” object.</p>
<p> </p>
<div class="section level4">
<h4 id="example-3-building-and-manipulating-dendrograms">Example 3: Building and manipulating dendrograms<a class="anchor" aria-label="anchor" href="#example-3-building-and-manipulating-dendrograms"></a>
</h4>
<p>The simple three-leaf dendrogram in Figure 1 can be created manually as follows:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## attach "leaf" and "label" attributes to leaf nodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"leaf"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"leaf"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"leaf"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"label"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"A"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"label"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"B"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"label"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"C"</span></span>
<span><span class="co">## set "height" attributes for all nodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"height"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"height"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"height"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"height"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"height"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="co">## set "midpoints" attributes for all nodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"midpoint"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.75</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"midpoint"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"midpoint"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"midpoint"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"midpoint"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="co">## set "members" attributes for all nodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"members"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"members"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"members"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">"members"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="st">"members"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="co">## set class as "dendrogram" </span></span>
<span><span class="co">## Note that setting the class for the root node</span></span>
<span><span class="co">## automatically sets the class of all nested subnodes</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"dendrogram"</span></span>
<span><span class="va">x</span></span>
<span><span class="co">#&gt; 'dendrogram' with 2 branches and 3 members total, at height 2</span></span></code></pre></div>
<p>As demonstrated above, manually setting attributes on dendrogram objects can be rather tedious, motivating the development of functions to automate the generation and manipulation of these tree structures.</p>
<p>This simple tree can be recreated more succinctly using the <strong>phylogram</strong> package functions as follows:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## recursively set class, midpoint, members and leaf attributes</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/remidpoint.html">remidpoint</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">## set incremental height attributes</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.cladogram.html">as.cladogram</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">## set label attributes using dendrapply</span></span>
<span><span class="va">set_label</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">node</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dendrogram.html" class="external-link">is.leaf</a></span><span class="op">(</span><span class="va">node</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">node</span>, <span class="st">"label"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">LETTERS</span><span class="op">[</span><span class="va">node</span><span class="op">]</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">node</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/dendrapply.html" class="external-link">dendrapply</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">set_label</span><span class="op">)</span></span>
<span><span class="va">x</span></span>
<span><span class="co">#&gt; 'dendrogram' with 2 branches and 3 members total, at height 2</span></span></code></pre></div>
<p>Similarly, dendrogram objects can be subset using either the <code>prune</code> function or standard list-subsetting syntax, again with the help of utility functions to recursively reset node attributes. The following code demonstrates one option for rearranging the tree with species A and B as sister taxa and C as the ancestor:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## isolate root node (species C)</span></span>
<span><span class="va">ancestor</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prune.html">prune</a></span><span class="op">(</span><span class="va">x</span>, pattern <span class="op">=</span> <span class="st">"C"</span>, keep <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co">## alternative option using subset operator</span></span>
<span><span class="va">ancestor</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">## create subtree without species C</span></span>
<span><span class="va">subtree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prune.html">prune</a></span><span class="op">(</span><span class="va">x</span>, pattern <span class="op">=</span> <span class="st">"C"</span><span class="op">)</span></span>
<span><span class="co">## graft subtree onto root</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ancestor</span>, <span class="va">subtree</span><span class="op">)</span></span>
<span><span class="co">## set attributes as above</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.cladogram.html">as.cladogram</a></span><span class="op">(</span><span class="fu"><a href="../reference/remidpoint.html">remidpoint</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## plot dendrogram</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">x</span>, yaxt <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figures/unnamed-chunk-9-1.png" width="500px" style="display: block; margin: auto auto auto 0;"></p>
<p>    <strong>Figure 3:</strong> Rearranged dendrogram with species C ancestral to A and B</p>
<p> </p>
</div>
</div>
<div class="section level3">
<h3 id="tree-visualization">Tree visualization<a class="anchor" aria-label="anchor" href="#tree-visualization"></a>
</h3>
<p>Publication-quality trees can be generated from dendrogram objects using the <strong>stats</strong> plotting function <code>plot.dendrogram</code>, and the extensive plotting functions available in dendrogram-enhancing packages such as <strong>circlize</strong> <span class="citation">(Gu et al. 2014)</span> and <strong>dendextend</strong> <span class="citation">(Galili 2015)</span>. The latter also offers the facility to convert dendrograms to “ggdend” objects, for which many powerful ‘grammar of graphics’ plotting functions are available in the <strong>ggplot2</strong> <span class="citation">(Wickham 2009)</span> and <strong>ggdendro</strong> <span class="citation">(DeVries and Ripley 2016)</span> packages. Moreover, there are several advanced plotting options for “phylo” objects in the <strong>ape</strong> package <span class="citation">(Paradis, Claude, and Strimmer 2004)</span>, as well as the Bioconductor package <strong>ggtree</strong> <span class="citation">(Yu et al. 2017)</span>. Given the extensive tree visualization options already available, we do not include any additional plotting functions in the <strong>phylogram</strong> package.</p>
<p> </p>
</div>
<div class="section level3">
<h3 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h3>
<p>The <strong>phylogram</strong> package offers a dendrogram parser for phylogenetic trees, several new tree-editing functions, and a bridge between the “dendrogram”” and “phylo” object types that improves accessibility to the comprehensive number of object-specific functions found across a suite of contributed packages. Future versions of the package will aim to further expand the range of input formats and object types available, thereby helping to integrate the wide variety of phylogenetic applications implemented in the R programming language. This software is still under active development, and will continue to be upgraded and expanded as new applications arise. Bug reports and other feedback are welcomed and can be directed to the GitHub issues page at <a href="http://github.com/ropensci/phylogram/issues" class="external-link uri">http://github.com/ropensci/phylogram/issues</a>, or the <strong>phylogram</strong> google group at <a href="https://groups.google.com/group/phylogram" class="external-link uri">https://groups.google.com/group/phylogram</a>.</p>
<p> </p>
</div>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>This software was developed with funding from a Rutherford Foundation Postdoctoral Research Fellowship from the Royal Society of New Zealand. The authors declare no competing interests.</p>
<p> </p>
</div>
<div class="section level2">
<h2 id="author-contributions">Author Contributions<a class="anchor" aria-label="anchor" href="#author-contributions"></a>
</h2>
<p>Both authors conceived and designed the software. SPW wrote the package functions and documentation. Both authors wrote the manuscript and gave final approval for publication.</p>
<p> </p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-deVries2016" class="csl-entry">
DeVries, Andrie, and Brian D Ripley. 2016. <em><span class="nocase">ggdendro: create dendrograms and tree diagrams using ’ggplot2’</span></em>. <a href="https://cran.r-project.org/package=ggdendro" class="external-link">https://cran.r-project.org/package=ggdendro</a>.
</div>
<div id="ref-Felsenstein1986" class="csl-entry">
Felsenstein, J, J Archie, WHE Day, W Maddison, C Meacham, FJ Rohlf, and D Swofford. 1986. <span>“<span class="nocase">The newick tree format, 1986</span>.”</span> <a href="http://evolution.genetics.washington.edu/phylip/newicktree.html" class="external-link">http://evolution.genetics.washington.edu/phylip/newicktree.html</a>.
</div>
<div id="ref-Galili2015" class="csl-entry">
Galili, Tal. 2015. <span>“<span class="nocase">dendextend: An R package for visualizing, adjusting and comparing trees of hierarchical clustering</span>.”</span> <em>Bioinformatics</em> 31 (22): 3718–20. <a href="https://doi.org/10.1093/bioinformatics/btv428" class="external-link">https://doi.org/10.1093/bioinformatics/btv428</a>.
</div>
<div id="ref-Gu2014" class="csl-entry">
Gu, Zuguang, Lei Gu, Roland Eils, Matthias Schlesner, and Benedikt Brors. 2014. <span>“<span class="nocase">circlize implements and enhances circular visualization in R</span>.”</span> <em>Bioinformatics</em> 30 (19): 2811–12. <a href="https://doi.org/10.1093/bioinformatics/btu393" class="external-link">https://doi.org/10.1093/bioinformatics/btu393</a>.
</div>
<div id="ref-Paradis2004" class="csl-entry">
Paradis, Emmanuel, Julien Claude, and Korbinian Strimmer. 2004. <span>“<span class="nocase">APE: analyses of phylogenetics and evolution in R language</span>.”</span> <em>Bioinformatics</em> 20: 289–90. <a href="https://doi.org/10.1093/bioinformatics/btg412" class="external-link">https://doi.org/10.1093/bioinformatics/btg412</a>.
</div>
<div id="ref-RCoreTeam2015" class="csl-entry">
R Core Team. 2017. <span>“<span class="nocase">R: A Language and Environment for Statistical Computing</span>.”</span> Vienna, Austria: R Foundation for Statistical Computing. <a href="https://cran.r-project.org/" class="external-link">https://cran.r-project.org/</a>.
</div>
<div id="ref-Revell2012" class="csl-entry">
Revell, Liam J. 2012. <span>“<span class="nocase">phytools: an R package for phylogenetic comparative biology (and other things)</span>.”</span> <em>Methods in Ecology and Evolution</em> 3 (2): 217–23. <a href="https://doi.org/10.1111/j.2041-210X.2011.00169.x" class="external-link">https://doi.org/10.1111/j.2041-210X.2011.00169.x</a>.
</div>
<div id="ref-Schliep2011" class="csl-entry">
Schliep, Klaus Peter. 2011. <span>“<span class="nocase">phangorn: phylogenetic analysis in R</span>.”</span> <em>Bioinformatics</em> 27 (4): 592–93. <a href="https://doi.org/10.1093/bioinformatics/btq706" class="external-link">https://doi.org/10.1093/bioinformatics/btq706</a>.
</div>
<div id="ref-Wickham2009" class="csl-entry">
Wickham, Hadley. 2009. <em><span class="nocase">ggplot2: Elegant Graphics for Data Analysis</span></em>. New York: Springer-Verlag. <a href="http://ggplot2.org" class="external-link">http://ggplot2.org</a>.
</div>
<div id="ref-Guangchuang2017" class="csl-entry">
Yu, Guangchuang, David Smith, Huachen Zhu, Yi Guan, and Tommy Tsan-Yuk Lam. 2017. <span>“<span class="nocase">ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data.</span>”</span> <em>Methods in Ecology and Evolution</em> 8: 28–36. <a href="https://doi.org/10.1111/2041-210X.12628" class="external-link">https://doi.org/10.1111/2041-210X.12628</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><!-- begin footer --><div class="footer">
    <div class="container">
        <div class="row start top-4 bottom-8">
            <div class="col-2"> <img id="footerlogo" src="https://ropensci.org/img/icon_short_white.svg">
</div>
            <div class="col-10">
                <div class="row">
                    <div class="col-md-4 col-xs-6">
                        <a href="https://github.com/ropensci" target="_blank" class="external-link"><div class="icon fab fa-github"></div></a>
                        <a href="https://github.com/ropenscilabs" target="_blank" class="external-link"><div class="icon fa fa-flask"></div></a>
                        <a href="https://hachyderm.io/@ropensci" target="_blank" class="external-link"><div class="icon fab fa-mastodon"></div></a>
                        <a href="https://vimeo.com/ropensci" target="_blank" class="external-link"><div class="icon fab fa-vimeo"></div></a>
                    </div>
                </div>
                <div class="row top-4">
                    <div class="col-md-2 col-sm-4">
                        <ul>
<h5 class="bottom-2">About</h5>
                            <li><a href="https://ropensci.org/about" class="external-link">About rOpenSci</a></li>
                            <li><a href="https://ropensci.org/software-review" class="external-link">Software Review</a></li>
                            <li><a href="https://ropensci.org/about#team" class="external-link">Our Team</a></li>
                            <li><a href="https://ropensci.org/careers" class="external-link">Jobs</a></li>
                            <li><a href="https://ropensci.org/donate" class="external-link">Donate</a></li>
                            <li><a href="https://ropensci.org/contact" class="external-link">Contact Us</a></li>
                        </ul>
</div>
                    <div class="col-md-3 col-sm-4">
                        <ul>
<h5 class="bottom-2">Community</h5>
                            <li><a href="https://ropensci.org/community/" class="external-link">Our Community</a></li>
                            <li><a href="https://ropensci.org/commcalls/" class="external-link">Community calls</a></li>
                            <li><a href="https://ropensci.org/events/" class="external-link">Events</a></li>
                            <li><a href="https://discuss.ropensci.org/" class="external-link">Join the Discussion</a></li>
                            <li><a href="https://ropensci.org/code-of-conduct" class="external-link">Code of conduct</a></li>
                        </ul>
</div>
                    <div class="col-md-2 col-sm-4">
                        <ul>
<h5 class="bottom-2">Resources</h5>
                            <li><a href="https://ropensci.org/packages/" class="external-link">Packages</a></li>
                            <li><a href="https://ropensci.org/usecases/" class="external-link">Use Cases</a></li>
                            <li><a href="https://ropensci.org/talks-papers/" class="external-link">Talks &amp; Publications</a></li>
                            <li><a href="https://docs.ropensci.org/" class="external-link">Documentation</a></li>
                            <li><a href="https://ropensci.org/news/" class="external-link">Newsletter</a></li>
                            <li><a href="https://ropensci.org/how-to-cite-ropensci/" class="external-link">Cite rOpenSci</a></li>
                        </ul>
</div>
                    <div class="col-md-4 col-xs-12">
                        <h5 class="bottom-2"></h5>

                        <p>rOpenSci is a fiscally sponsored project of <a href="http://numfocus.org" class="external-link">NumFOCUS</a>.</p>

                    </div>
                </div>
            </div>
        </div>
    </div>
</div>
<!-- / end footer -->


    </footer>
</div>

  

  

  </body>
</html>
